### DISCUSSION TITLE
[Bug] org.apache.dubbo.rpc.protocol.dubbo.FutureAdapter cannot be cast to org.apache.dubbo.samples.proto.GreetService

### DISCUSSION THREAD

**Discussion by Yoda-wu** [apache/dubbo#13914] [Question] [answered]
### Pre-check

- [X] I am sure that all the content I provide is in English.

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dubbo/issues?q=is%3Aissue) and found no similar issues.

### Apache Dubbo Component

Java SDK (apache/dubbo)

### Dubbo Version

operating system: ubuntu 22.04.1
dubbo-java-client :Dubbo: 3.3.0-beta.1; Java: 1.8
dubbo-go-server: Dubbo-go: 3.2.0; go 1.22.0

### Steps to reproduce this issue

bug_demo_repo: https://github.com/Yoda-wu/dubbo-go-samples/tree/bug_demo/service_discovery/service/go-server
1. start the nacos server
2. start the go-server:```go run server.go```
3. start the java-client: 
```
cd ../java-client 
mvn -e clean compile exec:java -Dexec.mainClass="org.apache.dubbo.samples.Main"

```
then:
![image](https://github.com/apache/dubbo/assets/64141194/4f5c5e4a-6def-4f24-bd65-e23de43252ce)

### What you expected to happen

i expect in java-client, it will receive `GreetResponse` from go-server

### Are you willing to submit a pull request to fix on your own?

- [X] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by AlbumenJ** [comment]
@EarthChen @icodening PTAL

**Comment by AlbumenJ** [comment]
@Yoda-wu Have you tried with 3.2.11?

**Reply by FoghostCn** [reply, marked as answer]
3.2.11 is OK

**Comment by FoghostCn** [comment]
https://github.com/apache/dubbo/blob/cf53c38179c13846cb760011758bd2035575fe4a/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistryDirectory.java#L776-L794

the `copiedInvocation` has different `returnType` with origin `invocation`,it should be `GreetResponse` but actually it is `CompletableFuture`

![image](https://github.com/apache/dubbo/assets/5020327/283a2a54-066e-401a-8e85-12d1dc58a84d)

finally return a `Future` to user:

https://github.com/apache/dubbo/blob/cf53c38179c13846cb760011758bd2035575fe4a/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/AsyncRpcResult.java#L238-L246

**Reply by AlbumenJ** [reply]
Can you please create a new issue based on this consequence and submit a PR to fix it?

**Reply by FoghostCn** [reply]
OK，I'll fix it soon

**Reply by FoghostCn** [reply]
I found it was already fixd in https://github.com/apache/dubbo/pull/13582 in 3.3.0-beta.2-SNAPSHOT

**Reply by FoghostCn** [reply]
There are still some problems, and errors will still be reported when using asynchronous methods.

```
CompletableFuture<GreetResponse> future = greetService.greetAsync(req);
GreetResponse greet = future.get();
```

```
Exception in thread "main" java.lang.ClassCastException: org.apache.dubbo.samples.proto.GreetResponse cannot be cast to java.util.concurrent.CompletableFuture
	at org.apache.dubbo.samples.proto.DubboGreetServiceTriple$GreetServiceStub.greetAsync(DubboGreetServiceTriple.java:117)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.dubbo.config.spring.util.LazyTargetInvocationHandler.invoke(LazyTargetInvocationHandler.java:54)
	at org.apache.dubbo.samples.proto.GreetServiceDubboProxy0.greetAsync(GreetServiceDubboProxy0.java)
	at org.apache.dubbo.samples.Main.main(Main.java:26)
```

code is at:
https://github.com/FoghostCn/dubbo-go-samples/blob/async-bug/service_discovery/service/java-client/src/main/java/org/apache/dubbo/samples/Main.java

I have opened a new issue https://github.com/apache/dubbo/issues/13931
